File: internal/rtk/orchestrator.go
package rtk

import "context"

type RunSessionOptions struct {
	Paths     Paths
	SpecPath  string
	SessionID string
	Force     bool
}

func failSession(paths Paths, session *Session, telemetryFile string, runErr error) (*Session, error) {
	if session.Status.State != "failed" {
		round := session.Status.Round
		if session.OpenRound != nil {
			round = session.OpenRound.Index
		}
		actor := ""
		if session.Status.ActiveActor != nil {
			actor = *session.Status.ActiveActor
		}
		phase := ""
		if session.Status.ActivePhase != nil {
			phase = *session.Status.ActivePhase
		}
		MarkSessionFailed(session, round, actor, phase, runErr)
		_ = SaveSession(paths, session)
	}
	_ = AppendTelemetryEvent(telemetryFile, map[string]any{
		"type":       "session_failed",
		"session_id": session.ID,
		"adapter":    session.Adapter,
		"round":      session.Status.Round,
		"error":      SanitizeError(runErr),
	})
	return nil, runErr
}

func RunSession(ctx context.Context, options RunSessionOptions) (*Session, error) {
	session, adapter, telemetryFile, err := bootstrapExecSession(options.Paths, options.SpecPath, options.SessionID, options.Force)
	if err != nil {
		return nil, err
	}
	_ = AppendTelemetryEvent(telemetryFile, map[string]any{
		"type":       "session_started",
		"session_id": session.ID,
		"adapter":    session.Adapter,
		"topic":      session.Topic,
		"chair":      session.Chair,
		"critics":    session.Critics,
		"max_rounds": session.MaxRounds,
	})

	evidenceKeyMap := map[string]string{}
	if err := runSeedPhase(ctx, adapter, session, evidenceKeyMap, telemetryFile, options.Paths); err != nil {
		return failSession(options.Paths, session, telemetryFile, err)
	}

	for round := 1; !roundLimitExceeded(session.MaxRounds, round); round++ {
		if err := runRound(ctx, adapter, session, round, evidenceKeyMap, telemetryFile, options.Paths); err != nil {
			return failSession(options.Paths, session, telemetryFile, err)
		}
		if session.Status.Converged {
			break
		}
	}

	if !session.Status.Converged && roundLimitExhausted(session.MaxRounds, session.Status.Round) {
		session.Status.State = "exhausted"
	}
	if err := SaveSession(options.Paths, session); err != nil {
		return failSession(options.Paths, session, telemetryFile, err)
	}
	_ = AppendTelemetryEvent(telemetryFile, map[string]any{
		"type":              "session_finished",
		"session_id":        session.ID,
		"adapter":           session.Adapter,
		"state":             session.Status.State,
		"round":             session.Status.Round,
		"converged":         session.Status.Converged,
		"unresolved_high":   session.Status.UnresolvedHigh,
		"unresolved_medium": session.Status.UnresolvedMedium,
	})
	return session, nil
}

File: src/orchestrator.js
const { appendEvidence, applyRound, createSession } = require("./kernel");
const { createAdapter } = require("./adapters");
const { createSessionFile, saveSession } = require("./store");

function collectEvidenceBatches(session, evidenceKeyMap, batches, round) {
  if (!Array.isArray(batches) || batches.length === 0) {
    return [];
  }

  const addedIds = [];
  for (const batch of batches) {
    if (!batch || !Array.isArray(batch.items) || batch.items.length === 0) {
      continue;
    }
    const added = appendEvidence(session, {
      items: batch.items,
      collectedBy: batch.collected_by,
      phase: batch.phase,
      round,
    });
    for (let index = 0; index < batch.items.length; index += 1) {
      const key = batch.items[index].key;
      if (!key) {
        continue;
      }
      if (evidenceKeyMap.has(key)) {
        throw new Error(`duplicate evidence key ${key}`);
      }
      evidenceKeyMap.set(key, added[index].id);
    }
    addedIds.push(...added.map((evidence) => evidence.id));
  }
  return addedIds;
}

async function runSession({ adapterKind, adapterConfig, sessionId, force = false }) {
  const adapter = createAdapter(adapterKind, adapterConfig);
  const metadata = adapter.metadata();

  const session = createSession({
    id: sessionId,
    topic: metadata.topic,
    chair: metadata.chair,
    critics: metadata.critics,
    max_rounds: metadata.max_rounds,
    adapter: adapter.kind,
  });
  createSessionFile(session, { force });

  const evidenceKeyMap = new Map();
  collectEvidenceBatches(session, evidenceKeyMap, await adapter.seedEvidence(), 0);
  saveSession(session);

  for (let round = 1; round <= session.max_rounds; round += 1) {
    const evidenceAdded = [];

    const explore = await adapter.collectEvidence({
      session,
      round,
      actor: session.chair,
      phase: "explore",
    });
    evidenceAdded.push(...collectEvidenceBatches(session, evidenceKeyMap, explore, round));

    const proposal = await adapter.propose({ session, round });

    const findings = [];
    for (const critic of session.critics) {
      const reExplore = await adapter.collectEvidence({
        session,
        round,
        actor: critic,
        phase: "re-explore",
      });
      evidenceAdded.push(...collectEvidenceBatches(session, evidenceKeyMap, reExplore, round));
      findings.push(...(await adapter.review({ session, round, critic, proposal, evidenceKeyMap })));
    }

    const verdict = await adapter.adjudicate({
      session,
      round,
      proposal,
      findings,
      evidenceKeyMap,
    });

    applyRound(session, {
      proposal,
      findings,
      verdict,
      evidence_added: evidenceAdded,
    });
    saveSession(session);

    if (session.status.converged) {
      break;
    }
  }

  if (!session.status.converged && session.status.round === session.max_rounds) {
    session.status = {
      ...session.status,
      state: "exhausted",
    };
  }
  saveSession(session);
  return session;
}

module.exports = {
  runSession,
};
